File: infra/maintenance/forge-token-audit.py
#!/usr/bin/env python3
"""Report drift between the declared forge-token inventory and the forge.

Reads `infra/platform/forge-tokens.toml`, the Forgejo SQLite database, and —
when configured — the credentials registry (`GET /api/credentials` on the jobs
API), and prints what disagrees. NEVER deletes a token and never reads one:
the hash and salt columns are excluded by name, and the only credential-shaped
value printed is `token_last_eight`, which is what Forgejo's own UI shows. The
registry side is values-free by construction — its rows carry storage
locations, never contents.

Exit codes follow the maintenance-check convention (see
infra/boss-audit-integrity-check.service): 0 clean, 2 drift found, 1 could not
run (including: a configured registry that could not be read — the run is
honest about being partial). Exit 2 makes the unit fail, which is what fires
the alert and leaves the maintenance Job open and loud.

WHY DRIFT AND NOT HEURISTICS. "Warn on tokens unused for 90 days" answers a
question nobody asked. The question that actually blocked a revocation on
2026-08-27 was "which of these may I delete", and no measurement of age can
answer it — only a statement of what SHOULD exist can. So the declaration is
the authority and this reports the difference, in both directions.

TWO DECLARATIONS, ONE COMPARISON EACH (CLAUDE.md 9a: this run is the equality
test that keeps them from silently disagreeing). forge-tokens.toml is the
forge-host inventory — per-token, hand-maintained, including the UNKNOWNs it
exists to name. The credentials registry is the platform's knowledge base —
per-credential, where a rotation-minted instance carries a packet-derived name
(`{id}-{first 8 of packet id}`), so registry matching is by id-prefix rather
than exact name. Both are compared against the live forge; a live token in
neither is doubly loud.

CAPABILITY, HONESTLY BOUNDED. The live-token side is readable credential-free
only on the forge host (the SQLite database). Reading it remotely would need
an admin-scoped forge token, which lives dispatcher-side as
BOSS_BROKER_FORGEJO_TOKEN and deliberately nowhere else — so run elsewhere,
this script says exactly what it cannot check rather than guessing.
"""

import argparse
import datetime
import json
import os
import sqlite3
import sys
import urllib.request

try:
    import tomllib  # 3.11+
except ModuleNotFoundError:  # pragma: no cover - exercised on older interpreters
    tomllib = None

SAFE_COLUMNS = ("id", "name", "scope", "created_unix", "updated_unix",
                "token_last_eight")
FORBIDDEN = {"token_hash", "token_salt"}


def when(v):
    if not v:
        return "never"
    return datetime.datetime.utcfromtimestamp(v).strftime("%Y-%m-%d %H:%M UTC")


def age_days(v, now):
    if not v:
        return None
    return int((now - v) / 86400)


def parse_declaration_minimal(text):
    """A deliberately tiny, STRICT reader for this one file's shape.

    `tomllib` only arrived in Python 3.11. The forge host runs 3.12 and the CI
    image is bookworm/3.11, so production never needs this — but a developer
    machine on 3.9 would otherwise be unable to run the script OR its tests,
    and a test that cannot run locally is one nobody runs before pushing.

    Handles exactly what forge-tokens.toml uses: `[[token]]` tables whose
    values are double-quoted single-line strings. Everything else RAISES rather
    than guessing, because a declaration silently half-parsed would report
    drift that is not there and, worse, miss drift that is.
    """
    tokens = []
    cur = None
    for lineno, raw in enumerate(text.splitlines(), 1):
        line = raw.strip()
        if not line or line.startswith("#"):
            continue
        if line == "[[token]]":
            cur = {}
            tokens.append(cur)
            continue
        if line.startswith("["):
            raise ValueError(
                "line %d: only [[token]] tables are understood, got %r" % (lineno, line))
        if cur is None:
            raise ValueError("line %d: key outside any [[token]] table: %r" % (lineno, line))
        key, sep, val = line.partition("=")
        if not sep:
            raise ValueError("line %d: expected `key = \"value\"`, got %r" % (lineno, line))
        key, val = key.strip(), val.strip()
        if len(val) < 2 or not val.startswith('"') or not val.endswith('"'):
            raise ValueError(
                "line %d: %r must be one double-quoted line; multi-line and bare "
                "values are not supported here" % (lineno, key))
        cur[key] = val[1:-1].replace('\\"', '"')
    return tokens


def load_declared(path):
    if tomllib is not None:
        with open(path, "rb") as fh:
            rows = tomllib.load(fh).get("token", [])
    else:
        with open(path, "r") as fh:
            rows = parse_declaration_minimal(fh.read())
    problems = []
    for i, t in enumerate(rows):
        for key in ("name", "consumer", "scope", "installed_at"):
            if not t.get(key):
                problems.append("declared token #%d is missing `%s`" % (i + 1, key))
    return rows, problems


def load_live(db_path, username):
    if not os.path.exists(db_path):
        raise FileNotFoundError(db_path)
    con = sqlite3.connect("file:%s?mode=ro" % db_path, uri=True)
    con.row_factory = sqlite3.Row
    have = {r["name"] for r in con.execute("PRAGMA table_info(access_token)")}
    cols = [c for c in SAFE_COLUMNS if c in have and c not in FORBIDDEN]
    user = con.execute(
        "SELECT id FROM user WHERE lower_name = ?", (username,)
    ).fetchone()
    if user is None:
        raise LookupError(username)
    rows = con.execute(
        "SELECT %s FROM access_token WHERE uid = ? ORDER BY created_unix"
        % ", ".join(cols),
        (user["id"],),
    ).fetchall()
    return [dict(r) for r in rows]


def audit(declared, live, stale_days, now):
    """Return (findings, unknown_count). A finding is (severity, text)."""
    findings = []
    by_name = {t["name"]: t for t in declared}
    live_names = {t["name"] for t in live}

    for t in live:
        name = t["name"]
        d = by_name.get(name)
        if d is None:
            findings.append((
                "UNDECLARED",
                "token %r (%s, ends %s) exists on the forge and is in no declaration. "
                "Either declare its consumer or delete it — an unattributable "
                "credential is one nobody dares revoke."
                % (name, t.get("scope") or "no scope", t.get("token_last_eight")),
            ))
            continue
        if (d.get("scope") or "") != (t.get("scope") or ""):
            findings.append((
                "SCOPE-DRIFT",
                "token %r is declared %r but the forge says %r"
                % (name, d.get("scope"), t.get("scope")),
            ))
        age = age_days(t.get("updated_unix"), now)
        if age is not None and age >= stale_days:
            findings.append((
                "STALE",
                "token %r last used %s (%d days ago), held by: %s"
                % (name, when(t.get("updated_unix")), age, d.get("consumer")),
            ))

    for d in declared:
        if d["name"] not in live_names:
            findings.append((
                "MISSING",
                "token %r is declared (consumer: %s) but does not exist on the "
                "forge. Either it was revoked without updating this file, or a "
                "consumer is about to fail." % (d["name"], d.get("consumer")),
            ))

    unknown = sum(
        1 for d in declared
        if d["name"] in live_names and str(d.get("consumer", "")).strip().upper() == "UNKNOWN"
    )
    return findings, unknown


def load_registry(registry_json, registry_url):
    """Return (rows, limit_message). rows is None when the registry was not
    read; limit_message says why when a configured source failed. Neither
    source needs a credential: the fixture is a file, and the jobs API's
    internal address trusts header-less callers for this read-only surface.
    """
    if registry_json:
        try:
            with open(registry_json) as fh:
                return json.load(fh), None
        except (OSError, ValueError) as e:
            return None, "cannot read registry fixture %s: %s" % (registry_json, e)
    if registry_url:
        url = registry_url.rstrip("/") + "/api/credentials"
        try:
            with urllib.request.urlopen(url, timeout=10) as resp:
                return json.loads(resp.read().decode("utf-8")), None
        except (OSError, ValueError) as e:
            # URLError (and its HTTPError subclass) are OSErrors; a
            # non-JSON body is a ValueError. Either way the direction
            # did not run, and saying so beats a silent skip.
            return None, "credentials registry at %s unreachable: %s" % (url, e)
    return None, None


def audit_registry(registry, live, user):
    """The registry direction, both ways. Returns (findings, skipped_rows).

    A live forge token belongs to a registry row when its name IS the row id
    or starts with `{id}-` — rotation-minted instances carry packet-derived
    names (`boss-dev-forge-token-7ee101aa`), and the row id is the durable
    identity they derive from. Rows whose principal does not mention the
    audited user are skipped (their tokens live under another account and
    this run cannot see them); the count is reported so the skip is visible.
    """
    findings = []

    def owns(row_id, name):
        return name == row_id or name.startswith(row_id + "-")

    forge_rows = [r for r in registry if r.get("kind") == "forgejo-access-token"]
    mine = [r for r in forge_rows if user in (r.get("principal") or "")]
    skipped = len(forge_rows) - len(mine)

    matched_ids = set()
    for t in live:
        name = t["name"]
        owners = [r for r in mine if owns(r["id"], name)]
        if not owners:
            findings.append((
                "REG-UNDECLARED",
                "token %r (%s, ends %s) exists on the forge and matches no "
                "credentials-registry row. Scope questions about it are "
                "experiments, not lookups — register it (GET /api/credentials "
                "is the reader) or delete it."
                % (name, t.get("scope") or "no scope", t.get("token_last_eight")),
            ))
            continue
        # Longest id wins: `boss-dev-forge-token-x` must credit the row
        # `boss-dev-forge-token`, not a hypothetical shorter prefix row.
        row = max(owners, key=lambda r: len(r["id"]))
        matched_ids.add(row["id"])
        declared_scopes = row.get("scopes") or []
        live_scopes = {s.strip() for s in (t.get("scope") or "").split(",") if s.strip()}
        if not declared_scopes:
            findings.append((
                "REG-SCOPE-UNVERIFIED",
                "registry row %r declares no scopes; the forge says its token %r "
                "carries %r — record that in the row, so the next scope question "
                "is a lookup." % (row["id"], name, t.get("scope") or ""),
            ))
        elif set(declared_scopes) != live_scopes:
            findings.append((
                "REG-SCOPE-DRIFT",
                "registry row %r declares scopes %s but forge token %r carries %r"
                % (row["id"], ",".join(sorted(declared_scopes)), name,
                   t.get("scope") or ""),
            ))

    for r in mine:
        if r["id"] not in matched_ids:
            findings.append((
                "REG-MISSING",
                "registry row %r (storage: %s) matches no live forge token for "
                "user %r — either the token was revoked without updating the "
                "registry, or the row does not record its forge-side token name "
                "yet." % (r["id"], r.get("storage_location"), user),
            ))
    return findings, skipped


def main(argv=None):
    p = argparse.ArgumentParser(description=__doc__)
    p.add_argument("--declaration", default="/opt/boss/infra/platform/forge-tokens.toml")
    p.add_argument("--db", default="/opt/forgejo/data/gitea/gitea.db")
    p.add_argument("--user", default="david")
    p.add_argument("--stale-days", type=int, default=30)
    p.add_argument("--now", type=int, default=None,
                   help="unix seconds; for tests, so the run is deterministic")
    p.add_argument("--registry-url", default=None,
                   help="jobs API base (e.g. http://10.20.0.34:7900); enables the "
                        "forge-vs-credentials-registry direction")
    p.add_argument("--registry-json", default=None,
                   help="a file holding the GET /api/credentials response; for "
                        "tests, and for a host that cannot reach the jobs API")
    args = p.parse_args(argv)

    now = args.now if args.now is not None else int(
        datetime.datetime.now(datetime.timezone.utc).timestamp())

    try:
        declared, decl_problems = load_declared(args.declaration)
    except OSError as e:
        print("forge-token-audit: cannot read declaration: %s" % e)
        return 1
    try:
        live = load_live(args.db, args.user)
    except FileNotFoundError as e:
        print("forge-token-audit: no forge database at %s — the live-token side is "
              "readable credential-free only on the forge host. Reading it remotely "
              "needs an admin-scoped forge token, which lives dispatcher-side as "
              "BOSS_BROKER_FORGEJO_TOKEN and deliberately not here; nothing was "
              "checked." % e)
        return 1
    except LookupError as e:
        print("forge-token-audit: no user %r in the forge database" % str(e))
        return 1

    print("forge-token-audit: %d declared, %d live on the forge for user %r"
          % (len(declared), len(live), args.user))

    if decl_problems:
        for m in decl_problems:
            print("  MALFORMED  %s" % m)
        print("forge-token-audit: declaration is malformed; fix it before trusting a clean run")
        return 2

    findings, unknown = audit(declared, live, args.stale_days, now)

    registry, registry_limit = load_registry(args.registry_json, args.registry_url)
    limited = False
    if registry is not None:
        reg_findings, reg_skipped = audit_registry(registry, live, args.user)
        forgejo_rows = sum(1 for r in registry if r.get("kind") == "forgejo-access-token")
        print("forge-token-audit: credentials registry holds %d row(s), %d of kind "
              "forgejo-access-token" % (len(registry), forgejo_rows))
        if reg_skipped:
            print("  %-12s %d registry row(s) belong to other principals and were "
                  "not checked against user %r's tokens"
                  % ("REG-SKIPPED", reg_skipped, args.user))
        findings += reg_findings
    elif registry_limit:
        # A configured registry that could not be read: the run is
        # partial and must say so rather than passing as complete.
        print("  %-12s forge-vs-registry NOT checked: %s" % ("LIMIT", registry_limit))
        limited = True
    else:
        print("forge-token-audit: no --registry-url/--registry-json; "
              "forge-vs-credentials-registry direction not checked")

    for sev, text in sorted(findings):
        print("  %-12s %s" % (sev, text))

    if unknown:
        print("  %-12s %d declared token(s) still have consumer = UNKNOWN. These are "
              "live credentials nobody can attribute; that is the debt this audit "
              "exists to retire." % ("UNATTRIBUTED", unknown))

    write_repo = [t for t in live if "write:repository" in (t.get("scope") or "")]
    if len(write_repo) > 1:
        print("  %-12s %d tokens carry write:repository (%s). Each is a full push "
              "credential; the fewer that exist, the smaller a disclosure is."
              % ("BREADTH", len(write_repo), ", ".join(sorted(t["name"] for t in write_repo))))

    if findings or unknown:
        print("forge-token-audit: %d finding(s), %d unattributed" % (len(findings), unknown))
        return 2
    if limited:
        print("forge-token-audit: no drift found, but the run was PARTIAL (see LIMIT above)")
        return 1
    print("forge-token-audit: clean — the forge matches the declaration%s"
          % (" and the credentials registry" if registry is not None else ""))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: crates/core/boss-testing/tests/forge_token_audit.rs
//! `infra/maintenance/forge-token-audit.py` is RUN, not read.
//!
//! The script's whole value is a verdict — exit 0 clean, exit 2 drift — that a
//! systemd unit turns into an alert. A test that greps the source for the word
//! "UNDECLARED" would pass on a script that never reaches the branch. So each
//! case here builds a throwaway Forgejo database and a declaration, executes
//! the real script against them, and asserts on what it actually decided.
//!
//! One of these tests is not about drift at all: the audit reads a table whose
//! other columns are a credential hash and its salt, and
//! `no_credential_material_reaches_the_output` pins that neither ever reaches
//! stdout. That property is the reason this script is allowed to exist.

use std::path::{Path, PathBuf};
use std::process::Command;

fn repo_root() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("../../..")
        .canonicalize()
        .expect("repo root resolves")
}

fn script() -> PathBuf {
    repo_root().join("infra/maintenance/forge-token-audit.py")
}

/// A scratch directory per case, so cases cannot see each other's fixtures.
fn scratch(case: &str) -> PathBuf {
    let dir = std::env::temp_dir().join(format!("forge-token-audit-{case}"));
    let _ = std::fs::remove_dir_all(&dir);
    std::fs::create_dir_all(&dir).expect("scratch dir");
    dir
}

/// The secret-shaped values the fixture stores. If either ever appears in the
/// audit's output, the script is leaking the thing it exists to avoid touching.
const HASH: &str = "d3adb33fd3adb33fd3adb33fd3adb33fd3adb33fd3adb33fd3adb33fd3adb33f";
const SALT: &str = "s4ltysaltysalty";

/// Build a Forgejo-shaped SQLite database holding `tokens`, each
/// `(name, scope, last_used_unix)`.
fn forge_db(dir: &Path, tokens: &[(&str, &str, i64)]) -> PathBuf {
    let db = dir.join("gitea.db");
    let mut py = String::from(
        "import sqlite3,sys\n\
         con=sqlite3.connect(sys.argv[1])\n\
         con.execute('CREATE TABLE user (id INTEGER PRIMARY KEY, lower_name TEXT)')\n\
         con.execute('CREATE TABLE access_token (id INTEGER PRIMARY KEY, uid INTEGER, \
             name TEXT, token_hash TEXT, token_salt TEXT, token_last_eight TEXT, \
             created_unix INTEGER, updated_unix INTEGER, scope TEXT)')\n\
         con.execute(\"INSERT INTO user VALUES (1,'david')\")\n",
    );
    for (i, (name, scope, used)) in tokens.iter().enumerate() {
        py.push_str(&format!(
            "con.execute('INSERT INTO access_token VALUES (?,?,?,?,?,?,?,?,?)',\
             ({id},1,'{name}','{HASH}','{SALT}','ab12cd34',1000000,{used},'{scope}'))\n",
            id = i + 1,
        ));
    }
    py.push_str("con.commit()\n");

    let out = Command::new("python3")
        .arg("-c")
        .arg(&py)
        .arg(&db)
        .output()
        .expect("python3 builds the fixture database");
    assert!(
        out.status.success(),
        "fixture db failed: {}",
        String::from_utf8_lossy(&out.stderr)
    );
    db
}

/// Write a declaration holding `tokens`, each `(name, consumer, scope)`.
fn declaration(dir: &Path, tokens: &[(&str, &str, &str)]) -> PathBuf {
    let path = dir.join("forge-tokens.toml");
    let mut toml = String::new();
    for (name, consumer, scope) in tokens {
        toml.push_str(&format!(
            "[[token]]\nname = \"{name}\"\nconsumer = \"{consumer}\"\n\
             scope = \"{scope}\"\ninstalled_at = \"somewhere\"\n\n"
        ));
    }
    std::fs::write(&path, toml).expect("write declaration");
    path
}

/// Write a credentials-registry fixture — the `GET /api/credentials`
/// response shape — holding forgejo rows, each `(id, scopes)`.
fn registry(dir: &Path, rows: &[(&str, &[&str])]) -> PathBuf {
    let path = dir.join("credentials.json");
    let rows: Vec<serde_json::Value> = rows
        .iter()
        .map(|(id, scopes)| {
            serde_json::json!({
                "id": id,
                "kind": "forgejo-access-token",
                "issuer": "forgejo (10.20.0.15)",
                "principal": "user david",
                "scopes": scopes,
                "storage_location": format!("k8s Secret somewhere/{id}"),
                "consumers": [],
                "rotation_policy": "on-demand",
                "rotated_at": null,
                "notes": "",
            })
        })
        .collect();
    std::fs::write(&path, serde_json::to_string(&rows).unwrap()).expect("write registry");
    path
}

/// Run the audit. Returns (exit code, stdout).
fn run(db: &Path, decl: &Path, now: i64) -> (i32, String) {
    run_args(db, decl, now, &[])
}

/// Run the audit with extra flags (the registry direction).
fn run_args(db: &Path, decl: &Path, now: i64, extra: &[&std::ffi::OsStr]) -> (i32, String) {
    let out = Command::new("python3")
        .arg(script())
        .arg("--db")
        .arg(db)
        .arg("--declaration")
        .arg(decl)
        .arg("--now")
        .arg(now.to_string())
        .args(extra)
        .output()
        .expect("the audit script runs");
    (
        out.status.code().unwrap_or(-1),
        String::from_utf8_lossy(&out.stdout).to_string(),
    )
}

const NOW: i64 = 1_800_000_000;

#[test]
fn a_forge_matching_its_declaration_is_clean() {
    let dir = scratch("clean");
    let db = forge_db(&dir, &[("boss-gcp", "write:repository", NOW - 3600)]);
    let decl = declaration(&dir, &[("boss-gcp", "the conductor", "write:repository")]);

    let (code, out) = run(&db, &decl, NOW);
    assert_eq!(code, 0, "expected a clean exit, got:\n{out}");
    assert!(out.contains("clean"), "{out}");
}

#[test]
fn a_token_the_declaration_does_not_mention_is_reported_and_fails_the_run() {
    let dir = scratch("undeclared");
    let db = forge_db(
        &dir,
        &[
            ("boss-gcp", "write:repository", NOW - 3600),
            ("push-20260818", "write:repository", NOW - 3600),
        ],
    );
    let decl = declaration(&dir, &[("boss-gcp", "the conductor", "write:repository")]);

    let (code, out) = run(&db, &decl, NOW);
    assert_eq!(code, 2, "undeclared token must fail the run:\n{out}");
    assert!(out.contains("UNDECLARED"), "{out}");
    assert!(
        out.contains("push-20260818"),
        "the finding must NAME the token, or it cannot be acted on:\n{out}"
    );
}

#[test]
fn a_declared_token_missing_from_the_forge_is_reported() {
    // The direction that catches a revocation nobody wrote down - and warns
    // before the consumer that still holds it discovers the loss itself.
    let dir = scratch("missing");
    let db = forge_db(&dir, &[("boss-gcp", "write:repository", NOW - 3600)]);
    let decl = declaration(
        &dir,
        &[
            ("boss-gcp", "the conductor", "write:repository"),
            ("boss-dev-read", "the dev pod", "read:repository"),
        ],
    );

    let (code, out) = run(&db, &decl, NOW);
    assert_eq!(code, 2, "{out}");
    assert!(out.contains("MISSING"), "{out}");
    assert!(out.contains("boss-dev-read"), "{out}");
    assert!(
        out.contains("the dev pod"),
        "naming the consumer is the point - it says who is about to break:\n{out}"
    );
}

#[test]
fn a_scope_that_widened_behind_our_back_is_reported() {
    let dir = scratch("scope");
    let db = forge_db(&dir, &[("boss-dev-read", "write:repository", NOW - 3600)]);
    let decl = declaration(&dir, &[("boss-dev-read", "the dev pod", "read:repository")]);

    let (code, out) = run(&db, &decl, NOW);
    assert_eq!(code, 2, "{out}");
    assert!(out.contains("SCOPE-DRIFT"), "{out}");
}

#[test]
fn an_unattributed_consumer_is_counted_even_when_nothing_else_drifts() {
    // consumer = UNKNOWN is the debt itself, so a forge that matches the
    // declaration perfectly must still NOT come back clean while one exists.
    let dir = scratch("unknown");
    let db = forge_db(&dir, &[("cluster-registry", "write:package", NOW - 3600)]);
    let decl = declaration(&dir, &[("cluster-registry", "UNKNOWN", "write:package")]);

    let (code, out) = run(&db, &decl, NOW);
    assert_eq!(
        code, 2,
        "an unattributable live credential is not a clean state:\n{out}"
    );
    assert!(out.contains("UNATTRIBUTED"), "{out}");
}

// ---------------------------------------------------------------------------
// The credentials-registry direction (packet 7ee101aa, second leg):
// live forge tokens vs registry rows of kind forgejo-access-token,
// both ways. The registry fixture is the GET /api/credentials shape.
// ---------------------------------------------------------------------------

#[test]
fn a_forge_matching_the_registry_is_clean_and_prefix_matching_covers_rotation_names() {
    // The rotation-minted instance carries a packet-derived name
    // (`{id}-{packet8}`); the registry row holds the durable id. If
    // matching were exact-name, every rotated token would false-alarm.
    let dir = scratch("reg-clean");
    let db = forge_db(
        &dir,
        &[(
            "boss-dev-forge-token-7ee101aa",
            "write:repository",
            NOW - 3600,
        )],
    );
    let decl = declaration(
        &dir,
        &[(
            "boss-dev-forge-token-7ee101aa",
            "the dev pod",
            "write:repository",
        )],
    );
    let reg = registry(&dir, &[("boss-dev-forge-token", &["write:repository"])]);

    let (code, out) = run_args(&db, &decl, NOW, &["--registry-json".as_ref(), reg.as_ref()]);
    assert_eq!(code, 0, "expected clean:\n{out}");
    assert!(out.contains("credentials registry"), "{out}");
    assert!(
        !out.contains("REG-"),
        "no registry finding expected:\n{out}"
    );
}

#[test]
fn a_live_token_absent_from_the_registry_names_itself() {
    let dir = scratch("reg-undeclared");
    let db = forge_db(
        &dir,
        &[
            (
                "boss-dev-forge-token-7ee101aa",
                "write:repository",
                NOW - 3600,
            ),
            ("rogue-token", "write:package", NOW - 3600),
        ],
    );
    // The TOML declares both, so every finding here is the registry's.
    let decl = declaration(
        &dir,
        &[
            (
                "boss-dev-forge-token-7ee101aa",
                "the dev pod",
                "write:repository",
            ),
            ("rogue-token", "somebody", "write:package"),
        ],
    );
    let reg = registry(&dir, &[("boss-dev-forge-token", &["write:repository"])]);

    let (code, out) = run_args(&db, &decl, NOW, &["--registry-json".as_ref(), reg.as_ref()]);
    assert_eq!(code, 2, "{out}");
    assert!(out.contains("REG-UNDECLARED"), "{out}");
    assert!(
        out.contains("'rogue-token'"),
        "the finding must NAME the token:\n{out}"
    );
    assert_eq!(
        out.matches("REG-UNDECLARED").count(),
        1,
        "exactly one — the rotation-named instance belongs to its row:\n{out}"
    );
}

#[test]
fn a_registry_row_the_forge_does_not_know_is_reported() {
    // The broker-root seed ships exactly this way — storage known,
    // forge-side token name unrecorded — so the message must offer
    // both readings: revoked-without-updating, or name-not-recorded.
    let dir = scratch("reg-missing");
    let db = forge_db(&dir, &[("boss-gcp", "write:repository", NOW - 3600)]);
    let decl = declaration(&dir, &[("boss-gcp", "the conductor", "write:repository")]);
    let reg = registry(
        &dir,
        &[
            ("boss-gcp", &["write:repository"]),
            ("boss-credential-broker-root", &[]),
        ],
    );

    let (code, out) = run_args(&db, &decl, NOW, &["--registry-json".as_ref(), reg.as_ref()]);
    assert_eq!(code, 2, "{out}");
    assert!(out.contains("REG-MISSING"), "{out}");
    assert!(out.contains("boss-credential-broker-root"), "{out}");
    assert!(
        out.contains("does not record its forge-side token name"),
        "the honest second reading must be offered:\n{out}"
    );
}

#[test]
fn registry_scope_drift_and_unverified_scopes_are_reported() {
    let dir = scratch("reg-scope");
    let db = forge_db(
        &dir,
        &[
            ("drifted-token", "write:repository", NOW - 3600),
            ("unverified-token", "read:user", NOW - 3600),
        ],
    );
    let decl = declaration(
        &dir,
        &[
            ("drifted-token", "somebody", "write:repository"),
            ("unverified-token", "somebody", "read:user"),
        ],
    );
    let reg = registry(
        &dir,
        &[
            // The registry believes read-only; the forge says write.
            ("drifted-token", &["read:repository"]),
            // The seeded honest-gap shape: scopes empty, audit fills.
            ("unverified-token", &[]),
        ],
    );

    let (code, out) = run_args(&db, &decl, NOW, &["--registry-json".as_ref(), reg.as_ref()]);
    assert_eq!(code, 2, "{out}");
    assert!(out.contains("REG-SCOPE-DRIFT"), "{out}");
    assert!(
        out.contains("REG-SCOPE-UNVERIFIED"),
        "an empty scopes array is a gap to fill, not a pass:\n{out}"
    );
    assert!(
        out.contains("'read:user'"),
        "the fill-me finding must SAY what the forge says, so recording it \
         is a copy rather than a re-derivation:\n{out}"
    );
}

#[test]
fn an_unreachable_registry_makes_the_run_honestly_partial() {
    // Port 1 refuses fast on any host. An otherwise-clean run that
    // could not check a configured direction must not exit 0.
    let dir = scratch("reg-unreachable");
    let db = forge_db(&dir, &[("boss-gcp", "write:repository", NOW - 3600)]);
    let decl = declaration(&dir, &[("boss-gcp", "the conductor", "write:repository")]);

    let (code, out) = run_args(
        &db,
        &decl,
        NOW,
        &["--registry-url".as_ref(), "http://127.0.0.1:1".as_ref()],
    );
    assert_eq!(code, 1, "partial is not clean:\n{out}");
    assert!(out.contains("LIMIT"), "{out}");
    assert!(
        out.contains("NOT checked"),
        "the run must say exactly what it could not do:\n{out}"
    );
}

#[test]
fn no_credential_material_reaches_the_output() {
    // THE PROPERTY THAT LETS THIS SCRIPT EXIST. It reads a table whose other
    // columns are a credential hash and its salt. Every case above is run
    // again here against one output, because a leak in any branch is a leak.
    let dir = scratch("noleak");
    let db = forge_db(
        &dir,
        &[
            ("boss-gcp", "write:repository", NOW - 3600),
            ("stale-one", "write:package", NOW - 90 * 86400),
            ("undeclared-one", "read:repository", NOW - 3600),
        ],
    );
    let decl = declaration(
        &dir,
        &[
            ("boss-gcp", "the conductor", "write:repository"),
            ("stale-one", "UNKNOWN", "write:package"),
            ("gone-one", "somebody", "read:repository"),
        ],
    );

    let (_code, out) = run(&db, &decl, NOW);
    assert!(!out.contains(HASH), "the token hash reached stdout:\n{out}");
    assert!(!out.contains(SALT), "the token salt reached stdout:\n{out}");
    // Sanity: this fixture really did exercise the reporting paths.
    assert!(
        out.contains("UNDECLARED") && out.contains("MISSING"),
        "{out}"
    );
}
